tests/test_hypothesis_plugin.py

```diff
@@ -0,0 +1,81 @@
+import pytest
+from hypothesis import given, settings, strategies as st
+
+import pydantic
+from pydantic.networks import import_email_validator
+
+
+def gen_models():
+    class MiscModel(pydantic.BaseModel):
+        # Each of these models contains a few related fields; the idea is that
+        # if there's a bug we have neither too many fields to dig through nor
+        # too many models to read.
+        obj: pydantic.PyObject
+        color: pydantic.color.Color
+        json_any: pydantic.Json
+
+    class StringsModel(pydantic.BaseModel):
+        card: pydantic.PaymentCardNumber
+        secbytes: pydantic.SecretBytes
+        secstr: pydantic.SecretStr
+
+    class UUIDsModel(pydantic.BaseModel):
+        uuid1: pydantic.UUID1
+        uuid3: pydantic.UUID3
+        uuid4: pydantic.UUID4
+        uuid5: pydantic.UUID5
+
+    class IPvAnyAddress(pydantic.BaseModel):
+        address: pydantic.IPvAnyAddress
+
+    class IPvAnyInterface(pydantic.BaseModel):
+        interface: pydantic.IPvAnyInterface
+
+    class IPvAnyNetwork(pydantic.BaseModel):
+        network: pydantic.IPvAnyNetwork
+
+    class StrictNumbersModel(pydantic.BaseModel):
+        strictbool: pydantic.StrictBool
+        strictint: pydantic.StrictInt
+        strictfloat: pydantic.StrictFloat
+        strictstr: pydantic.StrictStr
+
+    class NumbersModel(pydantic.BaseModel):
+        posint: pydantic.PositiveInt
+        negint: pydantic.NegativeInt
+        posfloat: pydantic.PositiveFloat
+        negfloat: pydantic.NegativeFloat
+
+    yield from (
+        MiscModel,
+        StringsModel,
+        UUIDsModel,
+        IPvAnyAddress,
+        IPvAnyInterface,
+        IPvAnyNetwork,
+        StrictNumbersModel,
+        NumbersModel,
+    )
+
+    try:
+        import_email_validator()
+    except ImportError:
+        pass
+    else:
+
+        class EmailsModel(pydantic.BaseModel):
+            email: pydantic.EmailStr
+            name_email: pydantic.NameEmail
+
+        yield EmailsModel
+
+
+@pytest.mark.parametrize('model', gen_models())
+@settings(max_examples=20)
+@given(data=st.data())
+def test_can_construct_models_with_all_fields(data, model):
+    # We take successful creation of an instance to demonstrate that Hypothesis
+    # knows how to provide valid values for each field, so we don't need any
+    # additional assertions.
+    instance = data.draw(st.from_type(model))
+    assert isinstance(instance, model)
```

Comment:
I was reading these tests and was thrown off by this assert at first: the value of the test is actually in being able to create the instances without raising `ValidationError`s.

The comment does point this out in retrospect, but I took it to refer to the `isinstance` assertion instead of to the absence of a `ValidationError`. I would maybe clarify that in the comment, or remove the assert since it doesn't really add a lot of value on top of being able to create the instance successfully.

Reply:
I've split the checks and comments, so that reading top-to-bottom makes it clear what we're checking and why:

```python
def test_can_construct_models_with_all_fields(data, model):
    # The value of this test is to confirm that Hypothesis knows how to provide
    # valid values for each field - otherwise, this would raise ValidationError.
    instance = data.draw(st.from_type(model))

    # We additionally check that the instance really is of type `model`, because
    # an evil implementation could avoid ValidationError by means of e.g.
    # `st.register_type_strategy(model, st.none())`, skipping the constructor.
    assert isinstance(instance, model)
```

Reply:
Very clear, thank you!